rhizome/host/lib/json_rpc_client.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+require "json"
+require "socket"
+
+class JsonRpcError < StandardError
+  attr_reader :code
+
+  def initialize(message, code)
+    super(message)
+    @code = code
+  end
+end
+
+class JsonRpcClient
+  def initialize(socket_path, timeout = 5, response_size_limit = 1048576)
+    @socket_path = socket_path
+    @timeout = timeout
+    @response_size_limit = response_size_limit
+  end
+
+  def call(method, params = {})
+    # id is used to correlate the context between request and response.
+    # See https://www.jsonrpc.org/specification
+    id = rand(10000000)
+
+    payload = {
+      jsonrpc: "2.0",
+      method: method,
+      params: params,
+      id: id
+    }
+
+    unix_socket = UNIXSocket.new(@socket_path)
+    unix_socket.write_nonblock(payload.to_json + "\n")
+
+    response = JSON.parse(read_response(unix_socket))
+    if response["error"]
+      raise JsonRpcError.new(response["error"]["message"], response["error"]["code"])
+    end
+
+    unix_socket.close
+
+    response["result"]
+  end
+
+  def read_response(socket)
+    buffer = +""
+    start_time = Time.now
+
+    begin
+      # Use IO.select to wait for data with a timeout. Subtract elapsed time,
+      # since this can be called multiple times.
+      elapsed_time = Time.now - start_time
+      ready_sockets = IO.select([socket], nil, nil, @timeout - elapsed_time)
+
+      # If ready_sockets is nil, it means timeout occurred
+      unless ready_sockets
+        socket.close
+        raise "The request timed out after #{@timeout} seconds."
+      end
+
+      # Loop until the whole JSON response is received.
+      loop do
+        buffer << socket.read_nonblock(4096)
+        break if valid_json?(buffer)
```

Comment:
This seems odd, loop until the thing is valid? Not only expensive, but seems a bit unsound?

Reply:
was discussing over slack, can't say I'm a huge fan of how jsonrpc (or teh spdk interpretation of it) omits some kind of simpler framing methodology, but indeed...there is no framing methodology except full blown streaming JSON